[LMLayer] Add default Unicode Word Breaking

eddieantonio opened the pull request:
This adds word breaking based on https://unicode.org/reports/tr29/#Default_Word_Boundaries.

The implementation is maintained over here: https://github.com/eddieantonio/unicode-default-word-boundary however, I've copy-pasted the important parts here.

I have a few concerns before this gets merged:

 - ~~I don't know how to export a member called `default` from a TypeScript namespace. I have named the word breaker `uax29` for now.~~ I both _implemented_ and _declared_ a namespace in the same file, which works, but is a bit inelegant.
 - I'm not sure I'm doing the `<reference path=...>` stuff correctly. I placed the automatically generated file in its own file, but I think it creates a global called `WORD_BREAK_PROPERTY`???? @jahorton, please help!

keyman-server commented:
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/151935:id/kvk/index.html)

eddieantonio commented:
Uh... why is the new test that I added <https://build.palaso.org/viewLog.html?buildTypeId=Keyman_Common_LMLayer_TestPullRequests&buildId=151933&tab=testsInfo&pager.currentPage=4> working, and yet old test fail on CI? They pass on my laptop...?

eddieantonio commented:
Update: at least one aspect of these build failures can be explained by #1785

eddieantonio commented:
Latest build failures are caused by this:

```
[23:34:52]Step 3/5: Build Web + Embedded (Command Line) (47s)
[23:34:53][Step 3/5] Starting: C:\BuildAgent\temp\agentTmp\custom_script1410269481087843935.cmd
[23:34:53][Step 3/5] in directory: C:\BuildAgent\work\24f935566396273\keyman\web\source
[23:34:53][Step 3/5] Node.js + dependencies check
[23:35:07][Step 3/5] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules\fsevents):
[23:35:07][Step 3/5] npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
[23:35:07][Step 3/5] 
[23:35:07][Step 3/5] added 416 packages in 12.399s
[23:35:07][Step 3/5] 
[23:35:07][Step 3/5] Compiling the Language Modeling layer module...
[23:35:07][Step 3/5] Dependencies check
[23:35:18][Step 3/5] npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules\fsevents):
[23:35:18][Step 3/5] npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
[23:35:18][Step 3/5] 
[23:35:18][Step 3/5] added 347 packages in 9.699s
[23:35:19][Step 3/5] 
[23:35:19][Step 3/5] > keyman-lmlayer-prototype@0.20181001.0 tsc C:\BuildAgent\work\24f935566396273\keyman\common\predictive-text
[23:35:19][Step 3/5] > tsc "-p" "./worker/tsconfig.json"
[23:35:19][Step 3/5] 
[23:35:23][Step 3/5] > wrap-worker-code LMLayerWorkerCode build/intermediate/index.js > build/intermediate/embedded_worker.js
[23:35:24][Step 3/5] 
[23:35:24][Step 3/5] > keyman-lmlayer-prototype@0.20181001.0 tsc C:\BuildAgent\work\24f935566396273\keyman\common\predictive-text
[23:35:24][Step 3/5] > tsc "-p" "./tsconfig.json"
[23:35:24][Step 3/5] 
[23:35:28][Step 3/5] Typescript compilation successful.
[23:35:28][Step 3/5] Language Modeling layer compilation successful.
[23:35:28][Step 3/5] 
[23:35:28][Step 3/5] Compile KMEI/KMEA build 1557
[23:35:29][Step 3/5] 
[23:35:29][Step 3/5] > keyman@11.0.0 tsc C:\BuildAgent\work\24f935566396273\keyman\web
[23:35:29][Step 3/5] > tsc "-p" "source/tsconfig.embedded.json"
[23:35:29][Step 3/5] 
[23:35:36][Step 3/5] Embedded TypeScript compiled as ../intermediate/keyman.js
[23:35:36][Step 3/5] 
[23:35:36][Step 3/5] Copy resources to ../intermediate/ui, .../osk
[23:35:36][Step 3/5] Copy source to ../intermediate/src
[23:35:36][Step 3/5] 
[23:35:36][Step 3/5] KeymanWeb resources saved under ../intermediate
[23:35:36][Step 3/5] 
[23:35:36][Step 3/5] Copying unminified embedded build products to ../release/unminified/embedded.
[23:35:36][Step 3/5] 
[23:35:36][Step 3/5] Copy resources to ../release/unminified/embedded/resources/ui, .../osk
[23:35:36][Step 3/5] Copy source to ../release/unminified/embedded/resources/src
[23:35:36][Step 3/5] 
[23:35:36][Step 3/5] KeymanWeb resources saved under ../release/unminified/embedded/resources
[23:35:36][Step 3/5] 
[23:35:39][Step 3/5] ../intermediate/keyman.js:17720: 
[23:35:39][Step 3/5] Originally at:
[23:35:39][Step 3/5] ../../common/predictive-text/build/intermediate/embedded_worker.js:2208: ERROR - unreachable code
[23:35:39][Step 3/5]                     if (!true) return [3 /*break*/, 13];
[23:35:39][Step 3/5]                                ^^^^^^^^^^^^^^^^^^^^^^^^^
[23:35:39][Step 3/5] 
[23:35:39][Step 3/5] 1 error(s), 0 warning(s)
[23:35:39][Step 3/5] Build failed.
[23:35:39][Step 3/5]  [H [J
[23:35:39][Step 3/5] Process exited with code 1
[23:35:41][Step 3/5] Process exited with code 1 (Step: Build Web + Embedded (Command Line))
[23:35:40][Step 3/5] Step Build Web + Embedded (Command Line) failed
```

The embedded worker contains this line:

```
if (!true) return [3 /*break*/, 13];
```

I BELIEVE this was generated by the TypeScript compiler, trying to translate an infinite while-loop within a `function*` generator to ES5 code. The while loop in question is this:

```typescript
while ( /* N.B., breaks at rule WB2. */true) {
      // Shift all positions, one scalar value to the right.
      rightPos = lookaheadPos;
      lookaheadPos = positionAfter(lookaheadPos);
      // Shift all properties, one scalar value to the right.
      [lookbehind, left, right, lookahead] =
        [left, right, lookahead, wordbreakPropertyAt(lookaheadPos)];

       // Break at the start and end of text, unless the text is empty.
      // WB1: Break at start of text...
      if (left === 'sot') {
        yield rightPos;
        continue;
      }
      // WB2: Break at the end of text...
      if (right === 'eot') {
        yield rightPos;
        break; // Reached the end of the string. We're done!
      }

     // ... SNIP ...
}
```

https://github.com/keymanapp/keyman/pull/1784/files#diff-e3139216810339a122994d468c4a0811R107

Then something down the line does some static-analysis and is upset with this compiled output? Not sure what the best way to proceed with this is...

eddieantonio commented:
I fixed the issue above by making the look _not_ infinite, and having it instead be conditional on something that will never not be true, due to a `break` statement.

Also, this exports the word breaker named `default`, which is a reserved word. Reserved words in JavaScript/TypeScript are weird, because they are only reserved _sometimes_. Most property accesses, e.g. `foo.x` can work just fine if you use a reserved word as the name of the property, e.g., `foo.if`, `foo.else`, `foo.while`. So `wordBreakers.default` is a perfectly valid expression ¯\_(ツ)_/¯

jahorton commented:
>     * I'm not sure I'm doing the `<reference path=...>` stuff correctly. I placed the automatically generated file in its own file, but I think it creates a global called `WORD_BREAK_PROPERTY`???? @jahorton, please help!

If you're familiar with C/C++ development, `<reference path=...>` is literally an `#include` preprocessor directive with an implied `#pragma once`.  Basically, the TS compiler will auto-dump the referenced source file at that location in the file if it has not already been included for the active compilation target file.  So yes, this creates the global you mention.

If you want that global to instead be within the same namespace, just wrap that global within a namespace of the same name and be sure to mark it with an `export` - variables of the same namespace from different source files are invisible to each other without an `export` keyword.

jahorton commented:
> Also, this exports the word breaker named `default`, which is a reserved word. Reserved words in JavaScript/TypeScript are weird, because they are only reserved _sometimes_. Most property accesses, e.g. `foo.x` can work just fine if you use a reserved word as the name of the property, e.g., `foo.if`, `foo.else`, `foo.while`. So `wordBreakers.default` is a perfectly valid expression 🤷‍♂ 

This won't be the first thing in our codebase abusing that fact.  You're fine.  (See the `keyman.interface` object in KMW.)

jahorton commented on common/predictive-text/worker/word_breaking/default-word-breaking/index.ts:
```diff
@@ -0,0 +1,369 @@
+   *
+   * @param text Text to find word boundaries in.
+   */
+  function* findBoundaries(text: string): Iterable<number> {
```
`function*`?

eddieantonio replied:
It's a [generator function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*)

Instead of doing the conventional `var a = []; a.push(x)` this _yields_ values one by one. In this case, it yields the string indices before which is a word boundary. In my original implementation, I intended this to be a generator to get indices one by one as required, but since I use `Array.from()` to yield all values and map over that array, using a generator is not useful now.

I don't think this is the IE11 issue, because I was looking at TS's compile sources, and it actually removes the `function*` syntax, replacing it with a polyfill. e.g., take a look at `embedded_worker.js`.

jahorton replied:
Ah, okay.  I tried looking it up when I saw it, but trying to Google search for `function*` doesn't exactly give that result; thanks!

jahorton commented on common/predictive-text/worker/word_breaking/default-word-breaking/index.ts:
```diff
@@ -0,0 +1,369 @@
+   * Generator that yields every successive span from the the text.
+   * @param text Any valid USVString to segment.
+   */
+  function* findSpans(text: string): Iterable<Span> {
```
`function*`?

eddieantonio commented:
I've removed all generator `function*`s, in preference for working exclusively with arrays. This avoids TypeScript generating a polyfill.

Addtionally, I've removed support for the word boundary rule [WB3c][], which was implemented using a regular expression using the `//u` flag, [which is unsupported in IE 11](). WB3c prevents a word break from appearing in the middle of [emoji ZWJ sequences][] like 🧚🏽‍♂️. In practice, not even VSCode handles this situation properly!

[WB3c]: https://unicode.org/reports/tr29/#WB3c
[emoji-zwj-sequences]: https://www.unicode.org/Public/emoji/12.0/emoji-zwj-sequences.txt

eddieantonio commented:
Uh... I'm not sure why Chrome on Android 7.1.1 is say that the Mocha test `describe` is not a function in ALL of the LMLayer's integration tests.

```
Uncaught ReferenceError: describe is not defined
at cases/top-level-lmlayer.js:4:1

ReferenceError: describe is not defined
```

eddieantonio commented:
So I've shrunk down the final size of `keymanweb.js` by a bit. With both turning the array of objects into an array of arrays, and by turning the string constants to a const enum (compiled to number literals), I took the size of the final bundle from **298.7KiB** to **270.3KiB** — a reduction of 28.9 KiB or about 0.33 jQueries. Here's a handy table to compare sizes of the final `keymanweb.js`:

| Package               | Size     | Size in bytes | Size in jQueries | 
|-----------------------|---------:|--------------:|-----------------:|
| Without word breaker  | 248.0KiB | 253939 bytes  | 2.88 jQueries    |
| With original word breaker | 298.7KiB | 305865 bytes  | 3.47 jQueries    |
| Optimization: arrays    | 279.1KiB | 285847 bytes  | 3.24 jQueries    |
| **Optimization: arrays and enums** | **270.3KiB** | **276784 bytes**  | **3.14 jQueries**    |
| jquery-3.4.1.min.js   |  86.1KiB |  88145 bytes  | 1.00 jQuery      |
